webpush/static/webpush/webpush_serviceworker.js

```diff
@@ -1,20 +1,34 @@
-
+let url;
 // Register event listener for the 'push' event.
 self.addEventListener('push', function(event) {
   // Retrieve the textual payload from event.data (a PushMessageData object).
   // Other formats are supported (ArrayBuffer, Blob, JSON), check out the documentation
   // on https://developer.mozilla.org/en-US/docs/Web/API/PushMessageData.
-  var payload = event.data ? event.data.text() : {"head": "No Content", "Body": "No Content"},
+  let payload = event.data ? event.data.text() : {"head": "No Content", "body": "No Content", "icon": ""},
     data = JSON.parse(payload),
     head = data.head,
-    body = data.body;
+    body = data.body,
+    icon = data.icon;
+    // If no url was received, it opens the home page of the website that sent the notification
+    // Whitout this, it would open undefined or the service worker file.
+    url = data.url ? data.url: self.location.origin;
 
   // Keep the service worker alive until the notification is created.
   event.waitUntil(
     // Show a notification with title 'ServiceWorker Cookbook' and use the payload
     // as the body.
     self.registration.showNotification(head, {
-      body: body
+      body: body,
+      icon: icon,
     })
   );
-});
\ No newline at end of file
+});
+
+self.addEventListener('notificationclick', function (event) {
```

Comment:
I am suspecting this will open same URL for several notification in same instance. As the URL is stored outside the push handeler function scope.

You can check proper example in [MDN documentation](https://developer.mozilla.org/en-US/docs/Web/API/PushEvent)
![image](https://user-images.githubusercontent.com/7114151/39970095-871c5c44-5707-11e8-8c6a-2b92146c52f9.png)